```{index} Variables
```

# Variables


```{contents}
:local:
:depth: 2
```


Chapter 1 wrote counts directly into expressions. An office-supply invoice needs names for the customer, quantity, and price so the same program can describe different orders. A **variable** associates a name and type with a value that the program can read or replace. Its **type** determines the permitted values and operations.


```{index} declaration; initialization; literal; local variable
```

## Declare, Initialize, and Use a Variable



A **declaration** introduces a variable. **Initialization** gives it its first value. In `int quantity = 12;`, `int` is the type, `quantity` is the name, and `12` is an integer **literal**, a value written in the source. The `=` operator assigns the right-hand value to the variable on its left.

```text
type variableName = expression;
```

Declare one variable per line while you are learning so its purpose and initial value are easy to inspect. The following program stores facts about an order and then prints them. The price details come in Types; begin by identifying each name and the value assigned to it.


In [1]:
string customer = "Riverside Dental";
string item = "Printer paper";
int quantity = 12;

Console.WriteLine(customer);
Console.WriteLine(item);
Console.WriteLine(quantity);


Riverside Dental
Printer paper
12


`customer` and `item` store text, while `quantity` stores a whole-number count. A declaration is not an output instruction. The output appears only because the program calls `Console.WriteLine`.

**Predict:** Change `quantity` to 20. Which output line changes? The customer and item stay the same because those variables were not reassigned.


```{index} definite assignment; unassigned local variable
```

### A Local Variable Must Be Assigned Before It Is Read



You can separate declaration from assignment. The compiler requires a local variable to be definitely assigned before a statement reads it.


In [1]:
int quantity;
quantity = 12;
Console.WriteLine(quantity);


12


This intentionally invalid fragment is shown as text rather than as a runnable cell:

```text
int quantity;
Console.WriteLine(quantity); // Compile-time error: no assigned value.
```

Do not assume that an unassigned local `int` automatically contains zero. Fields and array elements have initialization rules that you will study later; those rules do not make the local fragment valid.


```{index} assignment; reassignment; state trace
```

## Assignment Changes State



**State** is the collection of values the program currently holds. To change a variable, evaluate the expression on the right of `=` using the current values, then store its result in the variable on the left. Reassignment omits the type because the name has already been declared.


In [1]:
int quantity = 12;
quantity = quantity + 3;
Console.WriteLine(quantity);
quantity = quantity - 2;
Console.WriteLine(quantity);


15
13


| Statement just executed | quantity | Reason |
| --- | ---: | --- |
| `int quantity = 12;` | 12 | Initial order |
| `quantity = quantity + 3;` | 15 | Add three cases |
| `quantity = quantity - 2;` | 13 | Remove two cases |

`quantity = quantity + 3` is an instruction to replace a value, not a mathematical claim that a number equals itself plus three. The old value is used to compute the new one. `==`, taught in Operators, asks whether two values are equal.


```{index} assignment order; dependent values
```

### A Calculated Value Is Not a Live Formula



Assignment computes a result at that moment. It does not keep a spreadsheet-style formula that updates when another variable changes. Trace the next program before running it.


In [1]:
int quantity = 12;
int unitsPerCase = 10;
int totalUnits = quantity * unitsPerCase;
quantity = 20;
Console.WriteLine(totalUnits);
totalUnits = quantity * unitsPerCase;
Console.WriteLine(totalUnits);


120
200


The first output is 120 because the stored total has not been recalculated. The second is 200 after the program evaluates the expression again. This is a common cause of stale business totals: the program updates an input but prints an earlier result.


In [ ]:
// ### Exercise: Trace an Order Revision
// Start with quantity 12 and unitsPerCase 10. Add three cases, then calculate and print totalUnits. Expected total: 150.
// Your code starts here.

// Your code ends here.


In [1]:
// ### Solution
int quantity = 12;
int unitsPerCase = 10;
quantity = quantity + 3;
int totalUnits = quantity * unitsPerCase;
Console.WriteLine(totalUnits);


150


```{index} identifier; camelCase; naming rules; keywords
```

## Names Communicate Meaning



An **identifier** is a name in source code. For ordinary names in this book, begin with a letter or underscore and continue with letters, digits, or underscores. C# permits a wider set of Unicode identifier characters, but simple names are easier to share. Spaces and hyphens are not part of ordinary identifiers. Names are case-sensitive: `quantity` and `Quantity` differ.

Use **camelCase** for local variables, as in `unitPrice` and `casesPerBox`. Prefer a name that states the meaning or unit: `workMinutes` is clearer than `n`, and `taxRate` is clearer than `tax` for a proportion such as 0.07. Naming convention is a readability choice; language rules determine whether a name is legal.

| Candidate | Evaluation |
| --- | --- |
| `unitsPerCase` | Legal and descriptive |
| `_count` | Legal, though unnecessary here |
| `2ndOrder` | Invalid starting digit |
| `unit-price` | Parsed as subtraction, not one variable name |
| `int` | Reserved keyword; choose a descriptive name instead |

**Keywords** have special language meanings, such as `int` and `string`. Some keywords are reserved and some depend on context. You do not need to memorize a keyword table to write these programs.


```{index} var; type inference; const
```

## Explicit Types, var, and Constants



`var` asks the compiler to infer a local variable's type from its initializer. It does not mean “any type,” and the type does not change when a new value is assigned. `var quantity = 12;` declares an `int`; assigning text to it later would fail. `var` requires an initializer from which a type can be inferred.[^1]


In [1]:
var quantity = 12;
var customer = "Riverside Dental";
Console.WriteLine(quantity.GetType());
Console.WriteLine(customer.GetType());


System.Int32
System.String


`GetType()` displays the .NET type name: `System.Int32` for `int` and `System.String` for `string`. Use explicit types when they clarify the data model, especially for money or when a literal's inferred type is easy to misread.

A **constant** declared with `const` has a compile-time value that cannot be reassigned. Use it when that restriction fits the requirement; a changing order quantity should remain a variable.


In [1]:
const int casesPerBox = 5;
int quantity = 12;
Console.WriteLine(quantity / casesPerBox);


2


```{index} local scope; block scope
```

### Scope Limits Where a Name Can Be Used



The **scope** of a local variable is the region of code where its name can be used. These top-level variables belong to the program's generated entry point. A variable declared inside a nested block is available there, rather than everywhere in the application.


In [1]:
int quantity = 12;
{
    int addedCases = 3;
    quantity = quantity + addedCases;
    Console.WriteLine(addedCases);
}
Console.WriteLine(quantity);


3
15


After the closing brace, the program can read `quantity`, but it cannot read `addedCases`. Local scope is not “global scope.” Fields belonging to classes are a later topic.

**Exit check:** Explain why a declaration, a reassignment, and a read are different operations. Then trace a revised order without running it and check your prediction.


```{rubric} Footnotes
```

[^1]: Microsoft's [implicitly typed local variables](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/implicitly-typed-local-variables) reference explains how `var` obtains a fixed type from its initializer.
